implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs

```diff
@@ -61,32 +60,41 @@ impl Processor for UdpListenProcessor {
     }
 
     async fn process(&mut self, ctx: &mut Self::Context) -> Result<bool> {
-        debug!("Waiting for incoming UDP datagram...");
+        debug!(
+            "Waiting for incoming UDP datagram on {}...",
+            self.local_addr
+        );
         let (mut msg, addr) = match self.stream.next().await {
             Some(res) => match res {
                 Ok((msg, addr)) => (msg, addr),
-                Err(_e) => {
-                    info!("Failed to read message from UDP socket.");
+                Err(e) => {
+                    warn!("Failed to read message from UDP socket: {:?}", e);
+
+                    // Inform router of error
+                    let _ = self
+                        .router_handle
+                        .sender_listener_error(self.local_addr)
+                        .await;
+
+                    // Tell our relay we are done
                     return Ok(false);
                 }
             },
             None => {
-                info!("No message read, keep socket alive.");
+                info!("No message read, will wait for next message.");
                 return Ok(true);
             }
         };
 
-        // Register peer addr with sender half
-        // TODO: should `register` be called for every TransportMessage received?
-        self.router_handle
-            .register(self.tx_addr.clone(), addr.to_string())
-            .await?;
-
-        msg.return_route.modify().prepend(UdpAddress::from(addr));
-
-        debug!("Message onward route: {}", msg.onward_route);
-        debug!("Message return route: {}", msg.return_route);
+        // Set return route to go directly to paired sender, skipping the UDP router
+        msg.return_route
```

Comment:
You can use the `route!` macro to make it a bit more readable:
```rust
msg.return_route = route![
  self.sender_addr.clone(),
  Address::new(UDP, addr.to_string()),
  msg.return_route
];
```

Reply:
Done.